### PR TITLE
`PointerManager` support multi pointer event

### PR DESCRIPTION
**Pull request by cptbtptpbcptdtptp**
-  Optimization code.
- Support for multiple pointers.

### CODE REVIEW

**Comment by cptbtptpbcptdtptp** [comment]
#937

**Comment by cptbtptpbcptdtptp** [comment]
image

File: packages/core/src/input/pointer/Pointer.ts [hunk 1]
```diff
@@ -10,13 +12,90 @@ export class Pointer {
...
   readonly id: number;
+  /** @internal */
+  _events: PointerEvent[] = [];
+
   /** The phase of pointer. */
   phase: PointerPhase = PointerPhase.Leave;
+  /** The button of pointer. */
+  button: PointerButton;
+  /** The buttons of pointer. */
+  buttons: number;
```
Comment (GuoLei1990) [inline packages/core/src/input/pointer/Pointer.ts:23]: The API you add `buttons` and `button` comments is not clear, developer can't understand what they are express!

**Review by GuoLei1990 — CHANGES_REQUESTED** [review]

File: packages/core/src/input/pointer/Pointer.ts [hunk 2]
```diff
@@ -10,13 +12,90 @@ export class Pointer {
...
   position: Vector2 = new Vector2();
+  /** The change of the pointer. */
+  movingDelta: Vector2 = new Vector2();
+  // @todo: 这里用 frameCount 还是 event 自带的 timestamp ?
+  // 必要性：开发者可能需要判断 pointer 产生的先后顺序来决定用最早的 or 最近的 pointer
+  // timestamp 的好处是毫秒级，绝对可以判断出 pointer 的时序，但是可能有兼容性问题
+  // frameCount 的含义是代表它在哪一帧生成，好处是引擎原生无需引入时间戳概念，且无兼容问题
+  // 但是同一帧生成的 pointer 无法区分先后顺序
+  /** The frameCount the pointer was generated */
+  frameCount: number = -1;
```
Comment (GuoLei1990) [inline packages/core/src/input/pointer/Pointer.ts:34]: 如果不考虑兼容性，我喜欢 `timestamp` 而且帧率开发者可自行处理，但是 time stamp 可能会永久丢失

File: packages/core/src/input/pointer/PointerManager.ts [hunk 3]
```diff
@@ -35,25 +32,17 @@ export class PointerManager implements IInput {
...
-
-  private _currentPressedEntity: Entity;
-  private _currentEnteredEntity: Entity;
+  _downList: DisorderedArray<PointerButton> = new DisorderedArray();
 
   private _engine: Engine;
   private _canvas: Canvas;
   private _htmlCanvas: HTMLCanvasElement;
-  private _nativeEvents: PointerEvent[] = [];
+  private _nativeEvents = [];
```
Comment (GuoLei1990) [inline packages/core/src/input/pointer/PointerManager.ts:42]: Why you delete type use any

**Review by GuoLei1990 — CHANGES_REQUESTED** [review]

File: packages/core/src/input/pointer/PointerManager.ts [hunk 4]
```diff
@@ -81,32 +68,40 @@ export class PointerManager implements IInput {
...
-          }
+    const { _pointers: pointers, _nativeEvents: nativeEvents } = this;
+    /** Clean up the pointer released in the previous frame. */
+    let length = pointers.length;
+    if (length > 0) {
+      for (let i = length - 1; i >= 0; i--) {
+        if (i !== length - 1) {
+          pointers[i] = pointers[length - 1];
+          --length;
         }
```
Comment (GuoLei1990) [inline packages/core/src/input/pointer/PointerManager.ts:79]: Only left the end elment?

File: packages/core/src/input/pointer/PointerManager.ts [hunk 5]
```diff
@@ -81,32 +68,40 @@ export class PointerManager implements IInput {
...
+      nativeEvents.length = 0;
+    }
+
+    /** Pointer handles its own events. */
+    length = pointers.length;
+    if (length > 0) {
+      const updatePointer = this._engine.physicsManager._initialized
+        ? this._updatePointerWithPhysics
+        : this._updatePointerWithoutPhysics;
+      const { clientWidth, clientHeight } = this._htmlCanvas;
```
Comment (GuoLei1990) [inline packages/core/src/input/pointer/PointerManager.ts:101]: Can merge  `this._updatePointerWithPhysics` and `this._updatePointerWithoutPhysics` logic?

File: packages/core/src/input/pointer/PointerManager.ts [hunk 6]
```diff
@@ -81,32 +68,40 @@ export class PointerManager implements IInput {
...
+      pointers.length = length;
+    }
+
+    /** Generate the pointer received for this frame. */
+    length = nativeEvents.length;
+    if (length > 0) {
+      this._buttons = nativeEvents[length - 1].buttons;
+      for (let i = 0; i < length; i++) {
+        const evt = nativeEvents[i];
+        this._getPointer(evt.pointerId, frameCount)?._events.push(evt);
```
Comment (GuoLei1990) [inline packages/core/src/input/pointer/PointerManager.ts:90]: `evt.pointerId` amybe not start from 0, before we optimizaion this

File: packages/core/src/input/pointer/PointerManager.ts [hunk 7]
```diff
@@ -158,257 +154,194 @@ export class PointerManager implements IInput {
...
     const { _pointers: pointers } = this;
-    let deleteCount = 0;
-    const totalCount = pointers.length;
-    for (let i = 0; i < totalCount; i++) {
-      if (pointers[i].phase === PointerPhase.Leave) {
-        deleteCount++;
-      } else {
-        if (deleteCount > 0) {
-          pointers[i - deleteCount] = pointers[i];
+    const index = this._pointerIDMap.indexOf(pointerId);
```
Comment (GuoLei1990) [inline packages/core/src/input/pointer/PointerManager.ts:172]: I guess you want to fixed the reference and pointerID

Comment (GuoLei1990) [inline packages/core/src/input/pointer/PointerManager.ts:172]: Use `this._pointerIDMap` to replace  before `_uniqueID`

File: packages/core/src/input/pointer/Pointer.ts (hunk 1 above)
Comment (cptbtptpbcptdtptp) [reply packages/core/src/input/pointer/Pointer.ts:23]: These two APIs are required. Taking the mouse as an example, the button represents which key (`Primary`, `Auxiliary`, or `Secondary`) triggers the event, and the buttons represent the `hold down key` of the cursor.

**Review by GuoLei1990 — APPROVED** [review]

**Pull request merged**